src/notebook.c

```diff
@@ -754,6 +775,237 @@ void notebook_remove_page(gint page_num)
+	{
+		doc = g_array_index(docs, GeanyDocument*, i);
+
+		if (doc->file_name == NULL)
+			dirname = GEANY_STRING_UNTITLED;
+		else
+		{
+			tmp_dirname = g_path_get_dirname(DOC_FILENAME(doc));
+			dirname = get_doc_folder(tmp_dirname);
+			g_free(tmp_dirname);
+		}
+
+		tmp_docs = g_hash_table_lookup(table, dirname);
+
+		if (tmp_docs == NULL)
+		{
+			if (doc->file_name == NULL)
+				dirname = g_strdup(dirname);
+
+			tmp_docs = g_array_new(TRUE, TRUE, sizeof(GeanyDocument*));
+			g_hash_table_insert(table, dirname, tmp_docs);
+			g_array_append_val(dir_names, dirname);
+		}
+		else
+			if (doc->file_name != NULL)
+				g_free(dirname);
+
+		g_array_append_val(tmp_docs, doc);
+	}
+
+	g_array_sort(dir_names, compare_path_strings);
+
+	for (i = 0; i < dir_names->len; ++i)
+	{
+		dirname = g_array_index(dir_names, gchar*, i);
+		g_assert(dirname != NULL);
+		tmp_docs = g_hash_table_lookup(table, dirname);
+		g_array_sort(tmp_docs, compare_docs_based_on_filename);
+
+		for (j = 0; j < tmp_docs->len; ++j)
+		{
+			g_assert(save_index < docs_len);
+			doc = g_array_index(tmp_docs, GeanyDocument*, j);
+			g_array_index(docs, GeanyDocument*, save_index++) = doc;
+		}
+
+		g_array_free(tmp_docs, TRUE);
+		g_free(dirname);
+	}
+
+	g_assert(save_index == docs_len);
+	g_array_free(dir_names, TRUE);
+	g_hash_table_destroy(table);
+}
+
+
+void notebook_sort_tabs(NotebookTabSortMethod method)
+{
+	GArray *docs;
+	GeanyDocument* doc;
+	GtkWidget *child;
+	guint i;
+
+	docs = g_array_new(FALSE, TRUE, sizeof(GeanyDocument*));
+
+	foreach_document(i)
+		g_array_append_val(docs, documents[i]);
+
+	if (method == NOTEBOOK_TAB_SORT_FILENAME)
+		g_array_sort(docs, compare_docs_based_on_filename);
+	else
+		sort_docs_based_on_pathname(docs);
+
+	for (i = 0; i < docs->len; ++i)
+	{
+		doc = g_array_index(docs, GeanyDocument*, i);
+		child = document_get_notebook_child(doc);
+		g_assert(child != NULL);
+		gtk_notebook_reorder_child(GTK_NOTEBOOK(main_widgets.notebook), child, i);
+	}
+
+	g_array_free(docs, TRUE);
+}
+
+
+void on_sort_tabs_filename_activate(GtkMenuItem *menuitem, gpointer user_data)
+{
+	notebook_sort_tabs(NOTEBOOK_TAB_SORT_FILENAME);
+}
+
+
+void on_sort_tabs_pathname_activate(GtkMenuItem *menuitem, gpointer user_data)
+{
+	notebook_sort_tabs(NOTEBOOK_TAB_SORT_PATHNAME);
+}
+
+
+static gboolean on_idle_auto_sort_tabs(gpointer user_data)
+{
+	if (interface_prefs.show_notebook_tabs)
+	{
+		if (interface_prefs.auto_sort_tabs_filename)
+			notebook_sort_tabs(NOTEBOOK_TAB_SORT_FILENAME);
+		else if (interface_prefs.auto_sort_tabs_pathname)
+			notebook_sort_tabs(NOTEBOOK_TAB_SORT_PATHNAME);
+	}
+
+	on_idle_auto_sort_hooked = FALSE;
+	return FALSE;
+}
+
+
+static gboolean on_document_open(GeanyDocument* doc)
+{
+	if (!on_idle_auto_sort_hooked && interface_prefs.show_notebook_tabs &&
+		(interface_prefs.auto_sort_tabs_filename || interface_prefs.auto_sort_tabs_pathname))
+	{
+		if (g_idle_add(on_idle_auto_sort_tabs, NULL) > 0)
+			on_idle_auto_sort_hooked = TRUE;
```

Comment:
wouldn't it be better to only sort the new document?  In case someone changed sorting, it would probably not be expected that opening a new file re-sorts everything.  It would also likely be cheaper, simply walking the list once finding the appropriate new position.